src/box/lua/error.cc

```diff
@@ -109,6 +110,76 @@ error_create_table_case_get_reason(lua_State *L, int index)
 	return reason != NULL ? reason : "";
 }
 
+/**
+ * Encode client error argument to error.
+ *
+ * Error is described by `record'. Argument position is `arg_index`.
+ * Argument itself is at `stack_index` on Lua stack.
+ */
+static void
+luaT_error_client_encode_arg(lua_State *L, struct error *error,
+			     int stack_index, int arg_index,
+			     const struct errcode_record *record)
+{
+	const struct errcode_field *field = &record->errfields[arg_index];
+	switch (field->type) {
+	case ERRCODE_FIELD_TYPE_CHAR: {
+		const char *s = luaL_checkstring(L, stack_index);
+		if (strlen(s) > 1)
+			luaL_error(L, "expected single char as %d argument",
+				   stack_index);
+		error_set_str(error, field->name, s);
+		break;
+	}
+	case ERRCODE_FIELD_TYPE_INT: {
+		int64_t i = luaL_checkint64(L, stack_index);
+		if (i < INT_MIN || i > INT_MAX)
+			luaL_error(L,
+				   "expected integer in range [%d, %d] as "
+				   "%d argument", INT_MIN, INT_MAX,
+				   stack_index);
+		error_set_int(error, field->name, i);
+		break;
+	}
+	case ERRCODE_FIELD_TYPE_UINT: {
+		uint64_t u = luaL_checkuint64(L, stack_index);
+		if (u > UINT_MAX)
+			luaL_error(L,
+				   "expected integer in range [0, %u] as "
+				   "%d argument", UINT_MAX, stack_index);
+		error_set_uint(error, field->name, u);
+		break;
+	}
+	case ERRCODE_FIELD_TYPE_LLONG: {
+		int64_t i = luaL_checkint64(L, stack_index);
+		error_set_int(error, field->name, i);
+		break;
+	}
+	case ERRCODE_FIELD_TYPE_ULLONG: {
+		uint64_t u = luaL_checkuint64(L, stack_index);
+		error_set_uint(error, field->name, u);
+		break;
+	}
+	case ERRCODE_FIELD_TYPE_STRING: {
+		const char *s = luaL_checkstring(L, stack_index);
+		error_set_str(error, field->name, s);
+		break;
+	}
+	case ERRCODE_FIELD_TYPE_MSGPACK:
+		luaT_error_payload_set(L, error, field->name, stack_index);
+		break;
+	case ERRCODE_FIELD_TYPE_TUPLE: {
+		/* This is only for type check. */
+		struct tuple *t = luaT_checktuple(L, stack_index);
+		(void)t;
+		luaT_error_payload_set(L, error, field->name, stack_index);
```

Comment:
Please consider removing the checks and using `luaT_error_payload_set` for all types.

Reply:
Done. Please take a look at the updated last commit.